ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+#include "coll_acoll_utils.h"
+
+static inline int log_sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype,
+                                     int rank, int dim, int size, int sg_size, int cur_base,
+                                     int sg_start, struct ompi_communicator_t *comm,
+                                     mca_coll_base_module_t *module, ompi_request_t **preq,
+                                     int *nreqs)
+{
+    int msb_pos, sub_rank, peer, err;
+    int i, mask;
+    int end_sg, end_peer;
+
+    end_sg = sg_start + sg_size - 1;
+    if (end_sg >= size) {
+        end_sg = size - 1;
+    }
+    end_peer = (end_sg - cur_base) % sg_size;
+    sub_rank = (rank - cur_base + sg_size) % sg_size;
+
+    msb_pos = opal_hibit(sub_rank, dim);
+    --dim;
+
+    /* Receive data from parent in the sg tree. */
+    if (sub_rank > 0) {
+        assert(msb_pos >= 0);
+        peer = (sub_rank & ~(1 << msb_pos));
+        if (peer > end_peer) {
+            peer = (((peer + cur_base - sg_start) % sg_size) + sg_start);
+        } else {
+            peer = peer + cur_base;
+        }
+
+        err = MCA_PML_CALL(
+            recv(buff, count, datatype, peer, MCA_COLL_BASE_TAG_ALLGATHER, comm, MPI_STATUS_IGNORE));
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+
+    for (i = msb_pos + 1, mask = 1 << i; i <= dim; ++i, mask <<= 1) {
+        peer = sub_rank | mask;
+        if (peer >= sg_size) {
+            continue;
+        }
+        if (peer >= end_peer) {
+            peer = (((peer + cur_base - sg_start) % sg_size) + sg_start);
+        } else {
+            peer = peer + cur_base;
+        }
+        /* Checks to ensure that the sends are limited to the necessary ones.
+           It also ensures 'preq' not exceeding the max allocated. */
+        if ((peer < size) && (peer != rank) && (peer != cur_base)) {
+            *nreqs = *nreqs + 1;
+            err = MCA_PML_CALL(isend(buff, count, datatype, peer, MCA_COLL_BASE_TAG_ALLGATHER,
+                                     MCA_PML_BASE_SEND_STANDARD, comm, preq++));
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+    }
+
+    return err;
+}
+
+static inline int lin_sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype,
+                                     int rank, int dim, int size, int sg_size, int cur_base,
+                                     int sg_start, struct ompi_communicator_t *comm,
+                                     mca_coll_base_module_t *module, ompi_request_t **preq,
+                                     int *nreqs)
+{
+    int peer;
+    int err;
+    int sg_end;
+
+    sg_end = sg_start + sg_size - 1;
+    if (sg_end >= size) {
+        sg_end = size - 1;
+    }
+
+    if (rank == cur_base) {
+        for (peer = sg_start; peer <= sg_end; peer++) {
+            if (peer == cur_base) {
+                continue;
+            }
```

Comment:
In general we prefer the trick to iterate peers starting from rank+1 and wrap around to rank-1.

Reply:
Ack, will modify.